<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Class  — PXDataset-class • rpx</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">rpx</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lgatto/rpx">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Class </h1>
    </div>

    
    <p>An S4 class to store and access information about ProteomeXchange (PX)
  data sets. Schema versions 1.0, 1.1 and 1.2 are supported and are
  documented on the PX code repository:
  https://code.google.com/p/proteomexchange/source/browse/schema/</p>
    

        
    <h2 class="hasAnchor" id="objects-from-the-class"><a class="anchor" href="#objects-from-the-class"></a>Objects from the Class</h2>

    <p>Objects can be created with the constructor <code>PXDataset</code>.</p>
    
    <h2 class="hasAnchor" id="slots"><a class="anchor" href="#slots"></a>Slots</h2>

    <p><dl class='dl-horizontal'>
    <dt><code>id</code>:</dt><dd>Object of class <code>&quot;character&quot;</code> storing the
      datasets unique identifier. Can be accessed with <code>pxid()</code>.</dd> 
    <dt><code>formatVersion</code>:</dt><dd>Object of class <code>&quot;character&quot;</code>
      storing the version of the ProteomeXchange schema.  </dd>
    <dt><code>Data</code>:</dt><dd>Object of class <code>&quot;XMLNode&quot;</code> storing the
      ProteomeXchange description as XML node tree. </dd>
  </dl></p>
    
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    <p><dl class='dl-horizontal'></p>
    <p><dt>pxfiles</dt><dd><code>signature(object = &quot;PXDataset&quot;)</code>: return a
      <code>character</code> of all available files. </dd></p>
    <p><dt>pxget</dt><dd><code>signature(object = &quot;PXDataset&quot;, list, force =
      FALSE, ...)</code>: downloads the files from the ProteomeXchange
      repository. If <code>list</code> is missing, the file to be downloaded
      can be selected from a menu. If <code>list = &quot;all&quot;</code>, all files are
      downloaded. The file names, as returned by <code>pxfiles</code> can also
      be used. Alternatively, a <code>logical</code> or <code>numeric</code> indices
      can be used. All files will be downloaded in the working
      directory. Unless <code>force</code> is set to <code>TRUE</code>, files are
      not downloaded if already present in the working directory.
      Additional parameters can be passed to <code>download.file</code>
      via <code>...</code>.  Invisibly returns the names of the downloaded
      files. </dd></p>
    <p><dt>pxid</dt><dd><code>signature(object = &quot;PXDataset&quot;)</code>: returns the
      unique ProteomeXchange identifier. </dd></p>
    <p><dt>pxref</dt><dd><code>signature(object = &quot;PXDataset&quot;)</code>: returns the
      reference associcated with the <code>object</code>. </dd></p>
    <p><dt>pxtax</dt><dd><code>signature(object = &quot;PXDataset&quot;)</code>: returns the
      scientic taxonomic name of <code>object</code>. </dd></p>
    <p><dt>pxurl</dt><dd><code>signature(object = &quot;PXDataset&quot;)</code>: returns the
      base url on the ProteomeXchange server where
      <code>pxfiles(object)</code> reside. </dd></p>
    <p><dt>show</dt><dd><code>signature(object = &quot;PXDataset&quot;)</code>: textual
      representation of <code>object</code>. </dd>
  </dl></p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Vizcaino J.A. et al. &#39;ProteomeXchange: globally co-ordinated
  proteomics data submission and dissemination&#39;, Nature Biotechnology
  2014, 32, 223 -- 226, doi:10.1038/nbt.2839.</p>
    <p>Source repository for the ProteomeXchange project:
  https://code.google.com/p/proteomexchange/</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>px</span> <span class='kw'>&lt;-</span> <span class='fu'>PXDataset</span>(<span class='st'>"PXD000001"</span>)
<span class='no'>px</span></div><div class='output co'>#&gt; Object of class &quot;PXDataset&quot;
#&gt;  Id: PXD000001 with 10 files
#&gt;  [1] &#39;F063721.dat&#39; ... [10] &#39;erwinia_carotovora.fasta&#39;
#&gt;  Use &#39;pxfiles(.)&#39; to see all files.</div><div class='input'><span class='fu'>pxtax</span>(<span class='no'>px</span>)</div><div class='output co'>#&gt; [1] &quot;Erwinia carotovora&quot;</div><div class='input'><span class='fu'>pxurl</span>(<span class='no'>px</span>)</div><div class='output co'>#&gt; [1] &quot;ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2012/03/PXD000001&quot;</div><div class='input'><span class='fu'>pxref</span>(<span class='no'>px</span>)</div><div class='output co'>#&gt; [1] &quot;Gatto L, Christoforou A. Using R and Bioconductor for proteomics data analysis. Biochim Biophys Acta. 2014 Jan;1844(1 Pt A):42-51. Review&quot;</div><div class='input'><span class='fu'>pxfiles</span>(<span class='no'>px</span>)</div><div class='output co'>#&gt;  [1] &quot;F063721.dat&quot;                                                         
#&gt;  [2] &quot;F063721.dat-mztab.txt&quot;                                               
#&gt;  [3] &quot;PRIDE_Exp_Complete_Ac_22134.xml.gz&quot;                                  
#&gt;  [4] &quot;PRIDE_Exp_mzData_Ac_22134.xml.gz&quot;                                    
#&gt;  [5] &quot;PXD000001_mztab.txt&quot;                                                 
#&gt;  [6] &quot;TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML&quot; 
#&gt;  [7] &quot;TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzXML&quot;
#&gt;  [8] &quot;TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01.mzXML&quot;         
#&gt;  [9] &quot;TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01.raw&quot;           
#&gt; [10] &quot;erwinia_carotovora.fasta&quot;                                            </div><div class='input'><span class='no'>fnm</span> <span class='kw'>&lt;-</span> <span class='fu'>pxget</span>(<span class='no'>px</span>, <span class='st'>"PXD000001_mztab.txt"</span>)</div><div class='output co'>#&gt; <span class='message'>Downloading 1 file</span></div><div class='input'><span class='fu'>library</span>(<span class='st'>"MSnbase"</span>)</div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocGenerics</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: parallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘BiocGenerics’</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:parallel’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
#&gt; <span class='message'>    clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
#&gt; <span class='message'>    parLapplyLB, parRapply, parSapply, parSapplyLB</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    IQR, mad, sd, var, xtabs</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyDuplicated, append, as.data.frame, cbind, colMeans, colnames,</span>
#&gt; <span class='message'>    colSums, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
#&gt; <span class='message'>    grepl, intersect, is.unsorted, lapply, lengths, Map, mapply, match,</span>
#&gt; <span class='message'>    mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
#&gt; <span class='message'>    rbind, Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,</span>
#&gt; <span class='message'>    table, tapply, union, unique, unsplit, which, which.max, which.min</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    &#39;browseVignettes()&#39;. To cite Bioconductor, see</span>
#&gt; <span class='message'>    &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: mzR</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Rcpp</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocParallel</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: ProtGenerics</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>This is MSnbase version 2.1.5 </span>
#&gt; <span class='message'>  Read &#39;?MSnbase&#39; and references therein for information</span>
#&gt; <span class='message'>  about the package and how to get started.</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: ‘MSnbase’</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:stats’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    smooth</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from ‘package:base’:</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    trimws</span></div><div class='input'><span class='fu'>readMzTabData</span>(<span class='no'>fnm</span>, <span class='st'>"PEP"</span>)</div><div class='output co'>#&gt; MSnSet (storageMode: lockedEnvironment)
#&gt; assayData: 2351 features, 0 samples 
#&gt;   element names: exprs 
#&gt; protocolData: none
#&gt; phenoData: none
#&gt; featureData
#&gt;   featureNames: FGNDAR DGVSVAR ... LPALLK (2351 total)
#&gt;   fvarLabels: sequence accession ... peptide_abundance_std_error_sub.6.
#&gt;     (33 total)
#&gt;   fvarMetadata: labelDescription
#&gt; experimentData: use &#39;experimentData(object)&#39;
#&gt; Annotation:  
#&gt; - - - Processing information - - -
#&gt;  MSnbase version: 2.1.5 </div><div class='input'><span class='fu'>unlink</span>(<span class='st'>"PXD000001_mztab.txt"</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#objects-from-the-class">Objects from the Class</a></li>

      <li><a href="#slots">Slots</a></li>

      <li><a href="#methods">Methods</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
  Laurent Gatto &lt;lg390@cam.ac.uk&gt;

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
